libs/community/langchain_community/vectorstores/cratedb/base.py

```diff
@@ -0,0 +1,453 @@
+            env_key="CRATEDB_CONNECTION_STRING",
+        )
+
+        if not connection_string:
+            raise ValueError(
+                "Database connection string is required."
+                "Either pass it as a parameter, or set the "
+                "CRATEDB_CONNECTION_STRING environment variable."
+            )
+
+        return connection_string
+
+    @classmethod
+    def connection_string_from_db_params(
+        cls,
+        driver: str,
+        host: str,
+        port: int,
+        database: str,
+        user: str,
+        password: str,
+    ) -> str:
+        """Return connection string from database parameters."""
+        return str(
+            sqlalchemy.URL.create(
+                drivername=driver,
+                host=host,
+                port=port,
+                username=user,
+                password=password,
+                query={"schema": database},
+            )
+        )
+
+    def _select_relevance_score_fn(self) -> Callable[[float], float]:
+        """
+        The 'correct' relevance function
+        may differ depending on a few things, including:
+        - the distance / similarity metric used by the VectorStore
+        - the scale of your embeddings (OpenAI's are unit normed. Many others are not!)
+        - embedding dimensionality
+        - etc.
+        """
+        if self.override_relevance_score_fn is not None:
+            return self.override_relevance_score_fn
+
+        # Default strategy is to rely on distance strategy provided
+        # in vectorstore constructor
+        if self._distance_strategy == DistanceStrategy.COSINE:
+            return self._cosine_relevance_score_fn
+        elif self._distance_strategy == DistanceStrategy.EUCLIDEAN:
+            return self._euclidean_relevance_score_fn
+        elif self._distance_strategy == DistanceStrategy.MAX_INNER_PRODUCT:
+            return self._max_inner_product_relevance_score_fn
+        else:
+            raise ValueError(
+                "No supported normalization function for distance_strategy of "
+                f"{self._distance_strategy}. Consider providing relevance_score_fn to "
+                "CrateDBVectorSearch constructor."
+            )
+
+    @staticmethod
+    def _euclidean_relevance_score_fn(score: float) -> float:
+        """Return a similarity score on a scale [0, 1]."""
+        # The 'correct' relevance function
+        # may differ depending on a few things, including:
+        # - the distance / similarity metric used by the VectorStore
+        # - the scale of your embeddings (OpenAI's are unit normed. Many
+        #  others are not!)
+        # - embedding dimensionality
+        # - etc.
+        # This function converts the euclidean norm of normalized embeddings
+        # (0 is most similar, sqrt(2) most dissimilar)
+        # to a similarity function (0 to 1)
+
+        # Original:
+        # return 1.0 - distance / math.sqrt(2)
+        return score / math.sqrt(2)
```

Comment:
Please review this.

/cc @ckurze, @kneth, @surister

Reply:
Are those matters relevant and applicable here?

- https://github.com/crate/crate/issues/15835
- https://github.com/crate/crate/pull/15832

Reply:
Isn't `_score` based on BM25? If so,  is `1 - score` or `1 / score` closer to the API documentation?

Reply:
> Isn't `_score` based on BM25?

I think it is some sort of hybrid. @surister or @matriv may be able to elaborate further?

> If so, is `1 - score` or `1 / score` closer to the API documentation?

I also would like to defer this question to subject matter experts. ;]

Reply:
If someone wants to look into this, this guideline may help to exercise relevant code paths that leverage the `_euclidean_relevance_score_fn()` method.

It is those two test cases that will:
```python
tests/integration_tests/vectorstores/test_cratedb.py::test_cratedb_relevance_score
tests/integration_tests/vectorstores/test_cratedb.py::test_cratedb_retriever_search_threshold
```

You can invoke them on the spot, exclusively, by using this pytest command:
```shell
pytest -vvv tests/integration_tests/vectorstores/test_cratedb.py -k "test_cratedb_relevance_score or test_cratedb_retriever_search_threshold"
```

Reply:
Our full-text search function `match` gives a `_score` , which is the result of lucene's implementation of bm25. We also have `knn_match` and `vector_similarity` giving us a `_score`, even though then have they same name, they are different things.

Reply:
Is the goal here to return score normalized to a unit interval (0,1] where score itself based on euclidean distance?

As @surister mentioned, _score in fulltext is a different thing.

In CrateDB, both `knn_match` and `vector_similarity` use Lucene's `VectorSimilarityFunction.EUCLIDEAN` 
which is computed as `1 / (1 + squareDistance(v1, v2))`

We introduced  `vector_similarity` to get "pure" similarity value since `knn_match` can sometimes return _score outside of (0,1] interval, see https://github.com/crate-workbench/langchain/issues/19. It happens because Lucene's engine may take into account some other things when comparing docs which affect the final score.


1. I think parameter here must be called `distance`, at least in another repo `_euclidean_relevance_score_fn` 
   looks different https://github.com/crate-workbench/langchain/blob/cratedb/libs/core/langchain_core/vectorstores/base.py#L409-L421. 
   
 If physical meaning here is indeed score and not the distance, then it's either normalized and we don't need to do anything or it's affected by something else like in  https://github.com/crate-workbench/langchain/issues/19 and we can't divide by sqrt(2) as we don't know exactly what happened to the original score.

2. In case we use https://github.com/crate-workbench/langchain/blob/cratedb/libs/core/langchain_core/vectorstores/base.py#L409-L421. - it looks unaligned with CrateDB's `vector_similarity` both in docs and implementation.

Reply:
Hi. Thanks for your suggestions. I've responded to them [here](https://github.com/langchain-ai/langchain/pull/27710#discussion_r1832460771).

Reply:
Based on your suggestions, the patch has been adjusted with 7ff13192b1f. Thanks!